[JENKINS-68933] Better WebSocket testing, removal of reflection

jglick opened the pull request:
See [JENKINS-68933](https://issues.jenkins.io/browse/JENKINS-68933) and https://github.com/jenkinsci/jenkins/pull/6694#issuecomment-1174792020. Amending #4369.

### Proposed changelog entries

* N/A (internal changes only)

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).


<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6780"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

jglick committed:
[JENKINS-68933] Verify basic WebSocket connectivity in `RealJenkinsRule`

jglick commented:
Looking over https://github.com/eclipse/jetty.project/pull/6623 & https://github.com/eclipse/jetty.project/pull/3544 & https://github.com/eclipse/jetty.project/pull/5874.

jglick commented:
Rebased the test on top of the parent of #6694, where it passed.

jglick commented:
Tests appear to fail here because the build is running PR-merge strategy. @timja @dduportal I thought we had decided to switch builds of this repo to use PR-head strategy? Maybe it was only `bom`?

timja commented:
> Tests appear to fail here because the build is running PR-merge strategy. @timja @dduportal I thought we had decided to switch builds of this repo to use PR-head strategy? Maybe it was only `bom`?

only bom was switched

jglick committed:
Removing use of reflection to access Jetty WebSocket types

jglick committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins into ws-JENKINS-68933

jglick committed:
Removing `javax.servlet-api` from BOM

jglick committed:
Tweaking packaging

jglick commented:
Split off #6785 to attempt (so far without success) to use this framework to support Jetty 10.

basil committed:
Checkstyle

basil committed:
SpotBugs

basil committed:
Enforcer

basil reviewed (approved):
Thanks Jesse, I had started working on a similar PR earlier today but you obviously got it done a lot faster. In the interest of saving time I took the liberty of pushing a few commits to your branch to fix static analysis issues. I also left a comment in https://github.com/jenkinsci/jenkins/pull/6785#discussion_r914305175 explaining how to get this to work with Jetty 10.

basil commented:
Well now there's a a `javadoc: error - No public or protected classes found to document` but once that is resolved I think this is good to go.

jglick commented on pom.xml:
```diff
@@ -96,7 +96,7 @@ THE SOFTWARE.
     <spotbugs.effort>Max</spotbugs.effort>
     <spotbugs.threshold>Medium</spotbugs.threshold>
-    <spotbugs.excludeFilterFile>${project.basedir}/../src/spotbugs/spotbugs-excludes.xml</spotbugs.excludeFilterFile>
+    <spotbugs.excludeFilterFile>${maven.multiModuleProjectDirectory}/src/spotbugs/spotbugs-excludes.xml</spotbugs.excludeFilterFile>
```
TIL

jglick commented:
> In the interest of saving time

Well, there is no rush on this one since it does not fix anything for users, it is just making it easier to work on this code.

> I took the liberty of pushing a few commits to your branch to fix static analysis issues.

Thanks! I will check the Javadoc issue.

jglick committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins into ws-JENKINS-68933

jglick committed:
Skip Javadoc on new modules

jglick commented on websocket/jetty9/src/main/java/jenkins/websocket/Jetty9Provider.java:
```diff
@@ -0,0 +1,149 @@
+        };
+    }
+
+    private static Object createWebSocket(ServletUpgradeRequest req, ServletUpgradeResponse resp) {
```
Note the `Object` in the signature. Javadoc is super helpful

> a websocket object to use

It is more or less through trial and error that I concluded (in #4369) that a `WebSocketListener` is expected; “normal usage” apparently means adding annotations to POJOs and just remembering what method signatures to use. I think this misconceived style is inherited from the `javax.websocket` API. (Which tries to make you declare WS-aware endpoints with more annotations, and fails to provide a simple Java API call to upgrade a connection, forcing us to rely on Jetty-specific APIs.)

jglick committed:
Treat `ClosedChannelException` specially.

Otherwise closing connections sometimes produces stack traces

```
WARNING	j.websocket.WebSocketSession#error: unhandled WebSocket service error
java.nio.channels.ClosedChannelException
	at org.eclipse.jetty.websocket.core.internal.WebSocketSessionState.onEof(WebSocketSessionState.java:169)
	at org.eclipse.jetty.websocket.core.internal.WebSocketCoreSession.onEof(WebSocketCoreSession.java:253)
	at org.eclipse.jetty.websocket.core.internal.WebSocketConnection.fillAndParse(WebSocketConnection.java:482)
	at org.eclipse.jetty.websocket.core.internal.WebSocketConnection.onFillable(WebSocketConnection.java:340)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:319)
	at …
```

or

```
WARNING	j.agents.WebSocketAgents$Session#error
java.nio.channels.ClosedChannelException
	at org.eclipse.jetty.websocket.core.internal.WebSocketSessionState.onEof(WebSocketSessionState.java:169)
	at org.eclipse.jetty.websocket.core.internal.WebSocketCoreSession.onEof(WebSocketCoreSession.java:253)
	at …
```

though I am not sure if that is actually new behavior.

basil commented:
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

jglick commented on bom/pom.xml:
```diff
@@ -296,6 +296,11 @@ THE SOFTWARE.
       </dependency>
+      <dependency>
+        <groupId>org.kohsuke.metainf-services</groupId>
+        <artifactId>metainf-services</artifactId>
```
Note that this will cause that library to be managed for plugin POMs (like `credentials` or `artifact-manager-s3`) for the first time. Seems like a good thing; I am not sure why it was not already in the core BOM. But if this is a problem for any reason, it could trivially be reverted, and the version fixed in the new module.

basil replied:
FWIW there is an exclusion for this library elsewhere in the build: https://github.com/jenkinsci/jenkins/blob/644a261e78e9d80e2442787c8418cbdf283ac7e1/war/pom.xml#L92-L96=

jglick commented on websocket/spi/pom.xml:
```diff
@@ -0,0 +1,70 @@
+    <dependency>
+      <groupId>javax.servlet</groupId>
+      <artifactId>javax.servlet-api</artifactId>
+      <version>3.1.0</version>
```
Deliberately not using BOM version here.

basil replied:
Yeah, I saw that and I agree with using 3.1.0 because that is the lowest common denominator for Jetty 9 and Jetty 10. But the downside is this is adding the JAR to `WEB-INF/lib/javax.servlet-api-3.1.0.jar` in the WAR, which I think is undesirable because it is not actually used at runtime; rather, Winstone's copy is used. So to fix this I think we should exclude it when we pull in `websocket-spi` in `core/pom.xml`, which already has a `provided` dependency on `servlet-api` that is aligned with Jetty's.

jglick committed:
Do not bundle `javax.servlet-api`: https://github.com/jenkinsci/jenkins/pull/6780#discussion_r915086353

basil commented on websocket/jetty9/src/main/java/jenkins/websocket/Jetty9Provider.java:
```diff
@@ -0,0 +1,149 @@
+
+    private static final String ATTR_LISTENER = Jetty9Provider.class.getName() + ".listener";
+
+    // TODO does not seem possible to use HttpServletRequest.get/setAttribute for this
+    private static final Map<Listener, Session> sessions = new WeakHashMap<>();
```
Is there any concern that a session might get dropped from the map while it is still active, resulting in `IllegalStateException("missing session")` later on? From a correctness perspective it seems better to explicitly remove the session from the map in both the successful termination and failed termination cases.

basil replied:
Never mind, as soon as I posted this I remembered that it's the keys of a `WeakHashMap` that are weak references, and the lifetime of the `Listener` key is tied to the `HttpServletRequest`. :thumbsup:

jglick committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins into ws-JENKINS-68933

jglick committed:
Do not propagate `winstone` dep to `test`: https://github.com/jenkinsci/jenkins/pull/6785#discussion_r915259132

olamy reviewed (approved):
oh yes Jetty 10 API for websocket has changed a lot (it's supposed to be more easy to use but still a big change...)
good work here! (Hopefully we will not have to keep this jetty9 too long?)

basil committed:
Merge remote-tracking branch 'origin/master' into ws-JENKINS-68933

basil committed:
Update IntelliJ encodings

jglick commented on websocket/jetty9/src/main/java/jenkins/websocket/Jetty9Provider.java:
```diff
@@ -0,0 +1,149 @@
+    private static final String ATTR_LISTENER = Jetty9Provider.class.getName() + ".listener";
+
+    // TODO does not seem possible to use HttpServletRequest.get/setAttribute for this
+    private static final Map<Listener, Session> sessions = new WeakHashMap<>();
```
regression: #7076
